config/package_resouce.go

```diff
@@ -14,26 +14,229 @@
+			p.policy.Deb.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.Deb.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Msi:
+		pr := p.GetMsi()
+		if !packages.MSIExecExists {
+			return nil, fmt.Errorf("cannot manage MSI package because msiexec does not exist on the system")
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.MSI.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.MSI.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Googet:
+		pr := p.GetGooget()
+		if !packages.GooGetExists {
+			return nil, fmt.Errorf("cannot manage GooGet package %q because googet does not exist on the system", pr.GetName())
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.GooGet.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.GooGet.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Yum:
+		pr := p.GetYum()
+		if !packages.YumExists {
+			return nil, fmt.Errorf("cannot manage Yum package %q because yum does not exist on the system", pr.GetName())
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.Yum.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.Yum.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Zypper_:
+		pr := p.GetZypper()
+		if !packages.ZypperExists {
+			return nil, fmt.Errorf("cannot manage Zypper package %q because zypper does not exist on the system", pr.GetName())
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.Zypper.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.Zypper.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Rpm:
+		pr := p.GetRpm()
+		if !packages.RPMExists {
+			return nil, fmt.Errorf("cannot manage RPM package because rpm does not exist on the system")
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.RPM.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.RPM.Remove = pr
+		}
+	default:
+		return nil, errors.New("SystemPackage field not set or references unknown package manager")
+	}
+
+	return &ManagedResources{Packages: []ManagedPackage{p.policy}}, nil
+}
+
+// TODO: implement a caching system for installed packages.
+var aptInstalled []packages.PkgInfo
```

Comment:
Would it be better to have a map here keyed by pkg name so that lookups will be faster. for example in needsInstall function.

Reply:
[]packages.PkgInfo is what the inventory lookup call returns. Once I implement the cache system I may have that convert to a map though and remove the need for the intermediary function.